### PR TITLE
fix: simplify custom typing in storage module

### PR DESCRIPTION
**Pull request by dciborow**
## Description

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

- [ ] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.
- [ ] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [ ] I have updated the examples in README with the latest module version number.

### CODE REVIEW

File: modules/storage/storage-account/main.bicep [hunk 1]
```diff
@@ -86,75 +86,54 @@ type cors = {
...
-  description: roleAssignmentDescription?
-  roleDefinitionIdOrName: roleDefinitionIdOrName?
-  principalIds: principalIds?
-  principalType: principalType?
-  resourceType: resourceType?
+  description: string?
+  roleDefinitionIdOrName: string?
+  principalIds: string[]?
+  principalType: string?
+  resourceType: string?
```
Comment (jeskew) [inline modules/storage/storage-account/main.bicep:134]: I'm not sure if there are any style guidelines for Bicep types yet, but I prefer the approach you take in this PR. All types are purely structural, so nothing would prevent you from assigning a `resourceType` param something expecting a `principalType`. They're both strings, so the assignment wouldn't generate any diagnostics.

File: modules/storage/storage-account/main.bicep [hunk 2]
```diff
@@ -86,75 +86,54 @@ type cors = {
...
-
-type immutableStorageWithVersioning = {
-  enabled: bool
+  defaultEncryptionScope: string?
+  denyEncryptionScopeOverride: string?
+  immutableStorageWithVersioning: {
+    enabled: bool
+  }?
+  @allowed['Blob', 'Container', 'None']
+  publicAccess: string?
```
Comment (jeskew) [inline modules/storage/storage-account/main.bicep:126]:
```suggestion
  publicAccess: ('Blob' | 'Container' | 'None')?
```

`@allowed` is I believe only permitted on `param` statements.

**Review by jeskew — APPROVED** [review]

File: modules/storage/storage-account/main.bicep (hunk 1 above)
Comment (dciborow) [reply modules/storage/storage-account/main.bicep:133]: I also tried to nest it MORE, but did not feel it was better.

```
@description('The properties of a storage account’s Blob service.')
type blobServiceProperties = {
  @description('The blob service properties for change feed events.')
  changeFeed:  {
    enabled: bool
    @minValue(1)
    @maxValue(146000)
    retentionInDays: int?
  }?

  containerDeleteRetentionPolicy: containerDeleteRetentionPolicy?

  @description('Specifies CORS rules for the Blob service. You can include up to five CorsRule elements in the request. If no CorsRule elements are included in the request body, all CORS rules will be deleted, and CORS will be disabled for the Blob service.')
  cors: {
    corsRules: [
      {
        allowedHeaders: [
          'string'
        ]
        allowedMethods: [
          'string'
        ]
        allowedOrigins: [
          'string'
        ]
        exposedHeaders: [
          'string'
        ]
        maxAgeInSeconds: int
      }
    ]
  }?
  deleteRetentionPolicy: deleteRetentionPolicy?
  isVersioningEnabled: isVersioningEnabled?
  lastAccessTimeTrackingPolicy: lastAccessTimeTrackingPolicy?
  restorePolicy: restorePolicy?
}

```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -177,6 +177,79 @@ Instead go with:
 - `./modules/storage.bicep`
 - `./modules/compute.bicep`
 
+## User-Defined Types (Preview)
+
+The repository has been setup to leverage the preview feature, [user-defined types](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/user-defined-data-types), on a per module basis.
+
+While the feature is in Preview, [to enable](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/user-defined-data-types#enable-the-preview-feature) include a `bicepconfig.json` in your module directory.
+
+`type` assignments should be declared and the end of the file, after the `outputs` are declarationed. Types should include descriptions, which can be copied from the Azure API documentation when wrapping existing RP objects. Nested `type` declartions do not need a description. `type` should be reserved for use in complex cases, and should not be used to in place of primative types.
```
Comment (dciborow) [inline CONTRIBUTING.md:186]:
```suggestion
`type` assignments should be declared to the end of the file, after the declaration of `outputs`. Types should include descriptions, which can be copied from the Azure API documentation when wrapping existing RP objects. Nested `type` declarations do not need a description. `type` should be reserved for use in complex cases and should not be used to in place of primitive types.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -177,6 +177,79 @@ Instead go with:
...
+  @maxValue(365)
+  days: int?
+  enabled: bool
+}
+```
+
+Avoid these patterns.
+
+```bicep
+type enabled = bool  // Avoid using types to wrap simple primatives
```
Comment (shenglol) [inline CONTRIBUTING.md:221]:
```suggestion
type enabled = bool  // Avoid setting aliases for primitive types
```

File: modules/storage/storage-account/main.bicep [hunk 5]
```diff
@@ -214,11 +96,108 @@ module storageRbac 'modules/rbac.bicep' = [for (roleAssignment, index) in roleAs
...
+  enabled: bool
+}
+
+@description('Specifies CORS rules for the Blob service. You can include up to five CorsRule elements in the request. If no CorsRule elements are included in the request body, all CORS rules will be deleted, and CORS will be disabled for the Blob service.')
+type cors = {
+  corsRules: [
+    {
+      allowedHeaders: [
+        'string'
+      ]
```
Comment (jeskew) [inline modules/storage/storage-account/main.bicep:143]: Should these be `string[]` instead? The current type declaration will only permit the value `['string']`.

**Pull request merged**